pyvirtualcam/native_macos/main.mm

```diff
@@ -0,0 +1,165 @@
+    if (sMachServer == nil) {
+        return;
+    }
+
+	  uint8_t *outData = (uint8_t *)data;
+	  if (linesize != (cam_output_width * 2)) {
+		    blog(LOG_ERROR,
+		         "unexpected frame->linesize (expected:%d actual:%d)",
+		         (int)(cam_output_width * 2), (int)linesize);
+        return;
+	  }
+
+	  CGFloat width = cam_output_width;
+	  CGFloat height = cam_output_height;
+
+	  [sMachServer sendFrameWithSize:NSMakeSize(width, height)
+			       timestamp:timestamp
+			    fpsNumerator:cam_fps_num
+			  fpsDenominator:cam_fps_den
+			      frameBytes:outData];
+}
+
+
+void start(int width, int height, double fps, int delay) {
+    if (!virtual_output_start(width, height, fps, delay))
+        throw std::runtime_error("error starting virtual camera output");
+}
+
+void send(py::array_t<uint8_t, py::array::c_style> frame) {
+    py::buffer_info buf = frame.request();
+    if (buf.ndim != 3)
+        throw std::runtime_error("ndim must be 3 (h,w,c)");
+    if (buf.shape[2] != 4)
+        throw std::runtime_error("frame must have 4 channels (rgba)");
+
+    // We must handle port messages, and somehow our RunLoop isn't normally active
+    NSRunLoop *runLoop;
+  	NSDate *timeout = [NSDate dateWithTimeIntervalSinceNow:0.001];
+    runLoop = [NSRunLoop currentRunLoop];
+    [runLoop runUntilDate:timeout];
+    
+    uint64_t timestamp = mach_absolute_time();
+
+    uint8_t* data = (uint8_t*) malloc(buf.shape[1] * buf.shape[0] * 2);
+
+    // Convert RGBA to UYVY
+    for(int y = 0; y < buf.shape[0]; y++) {
+        for(int x = 0; x < buf.shape[1] / 2; x++) {
+            uint8_t* uyvy = &data[((y * buf.shape[1] + x * 2) * 2)];
+            const uint8_t* rgba = ((const uint8_t*)buf.ptr) + ((y * buf.shape[1] + x * 2) * 4);
+
+            // Downsample
+            uint8_t mixRGB[3];
+            mixRGB[0] = (rgba[0+0] + rgba[4+0]) / 2;
+            mixRGB[1] = (rgba[0+1] + rgba[4+1]) / 2;
+            mixRGB[2] = (rgba[0+2] + rgba[4+2]) / 2;
+
+            // Get U and V
+            uint8_t u;
+            uint8_t v;
+            UVfromRGB(&u, &v, mixRGB[0], mixRGB[1], mixRGB[2]);
+
+            // Variable for Y
+            uint8_t y;
+
+            // Pixel 1
+            YfromRGB(&y, rgba[0+0], rgba[0+1], rgba[0+2]);
+            uyvy[0] = u;
+            uyvy[1] = y;
+
+            // Pixel 2
+            YfromRGB(&y, rgba[4+0], rgba[4+1], rgba[4+2]);
+            uyvy[2] = v;
+            uyvy[3] = y;
+        }
+    }
+
+    virtualcam_output_raw_video(data, buf.shape[1] * 2, timestamp);
+    free(data);
+}
+
+void stop() {
+    virtual_output_stop();
+}
+
+
+PYBIND11_MODULE(_native_macos, m) {
+    m.def("start", &start, R"pbdoc(
+        Start the virtual cam output.
+    )pbdoc");
+
+    m.def("stop", &stop, R"pbdoc(
+        Stop the virtual cam output.
+    )pbdoc");
+
+    m.def("send", &send, R"pbdoc(
+        Send frame to the virtual cam.
+    )pbdoc");
+}
```

Comment:
README needs information that macOS is supported

Reply:
Opened https://github.com/letmaik/pyvirtualcam/issues/21.